File: java_realisation/src/main/java/com/lab2/java/gui/ListActionListenerOverride.java
package com.lab2.java.gui;

import com.lab2.java.doublyLinkedList.DoublyLinkedList;
import com.lab2.java.doublyLinkedList.LinkedList;
import com.lab2.java.doublyLinkedList.ListSerializeDeserialize;

import java.io.FileNotFoundException;

public class ListActionListenerOverride extends AbstractListActionListener {
    protected DoublyLinkedList<Object> items = new LinkedList<>();

    @Override
    public void onAdd(String text) {
        if (text.isEmpty()) return;
        Object data = builder.createFromString(text);
        items.addM(data);
        listModel.addElement(data);
    }

    @Override
    public void onInsert(String text, int index) {
        if (text.isEmpty()) return;
        Object data = builder.createFromString(text);
        items.addM(data, index);
        listModel.add(index, data);
    }

    @Override
    public void onRemove(int index) {
        items.removeM(index);
        listModel.remove(index);
    }

    @Override
    public void onSort() {
        items.sort(builder.getTypeComparator());
        listModel.clear();
        items.forEachM(listModel::addElement);
    }

    @Override
    public void onSave() {
        try {
            ListSerializeDeserialize.saveToFile(filename, items, builder);
        } catch (FileNotFoundException e) {
            System.err.println("Нельзя записать список в файл!");
            e.printStackTrace();
        }
    }

    @Override
    public void onLoad() {
        try {
            items = ListSerializeDeserialize.loadFromFile(filename, builder, new LinkedList<>());
            listModel.clear();
            items.forEachM(listModel::addElement);
        } catch (Exception e) {
            System.err.println("Нельзя прочитать список из файла!");
            e.printStackTrace();
        }
    }

    @Override
    public void onClearList() {
        items.removeAllM();
        listModel.removeAllElements();
    }
}

File: scala_realisation/src/main/scala/com/lab2/scala/gui/ListActionListenerOverride.scala
package com.lab2.scala.gui

import com.lab2.java.doublyLinkedList.DoublyLinkedList
import com.lab2.java.doublyLinkedList.LinkedList
import com.lab2.java.doublyLinkedList.ListSerializeDeserialize
import com.lab2.java.gui.AbstractListActionListener
import com.lab2.scala.doublyLinkedList.MyLinkedList

import java.io.FileNotFoundException

class ListActionListenerOverride extends AbstractListActionListener {

  var items: DoublyLinkedList[Object] = new LinkedList[Object]

  override def onAdd(text: String): Unit = {
    if (text.isEmpty) return
    val value = builder.createFromString(text)
    items.addM(value)
    listModel.addElement(value)
  }

  override def onInsert(text: String, index: Int): Unit = {
    if (text.isEmpty) return
    val value = builder.createFromString(text)
    items.addM(value, index)
    listModel.add(index, value)
  }

  override def onRemove(index: Int): Unit = {
    items.removeM(index)
    listModel.remove(index)
  }

  override def onSort(): Unit = {
    items.sort(builder.getTypeComparator)
    listModel.clear()
    items.forEachM(el => listModel.addElement(el))
  }

  override def onSave(): Unit = {

    try ListSerializeDeserialize.saveToFile(filename, items, builder)
    catch {
      case e: FileNotFoundException =>
        System.err.println("Невозможно записать список в файл!")
        e.printStackTrace()
    }
  }

  override def onLoad(): Unit = {
    try {
      items = ListSerializeDeserialize.loadFromFile(filename, builder, new LinkedList[AnyRef])
      listModel.clear()
      items.forEachM(el => listModel.addElement(el))
    } catch {
      case e: Exception =>
        System.err.println("Невозможно прочитоть список из файла!")
        e.printStackTrace()
    }
  }

  override def onClearList(): Unit = {
    items.removeAllM()
    listModel.removeAllElements()
  }
}
